Source/Core/AudioCommon/OpenALStream.cpp

```diff
@@ -2,70 +2,131 @@
+  p##func = (func##_t)::GetProcAddress(s_openal_dll, #func);                                       \
+  if (!p##func)                                                                                    \
+  {                                                                                                \
+    return false;                                                                                  \
+  }
 
-#ifdef _WIN32
-#pragma comment(lib, "openal32.lib")
-#endif
+OPENAL_API_VISIT(DYN_FUNC_DECLARE);
+
+static bool InitFunctions()
+{
+  OPENAL_API_VISIT(OPENAL_FUNC_LOAD);
+  return true;
+}
+
+static bool InitLibrary()
+{
+  if (s_openal_dll)
+    return true;
+
+  s_openal_dll = ::LoadLibrary(TEXT("openal32.dll"));
+  if (!s_openal_dll)
+    return false;
+
+  if (!InitFunctions())
+  {
+    ::FreeLibrary(s_openal_dll);
+    s_openal_dll = nullptr;
+    return false;
+  }
+
+  return true;
+}
+
+bool OpenALStream::isValid()
+{
+  return InitLibrary();
+}
 
 //
 // AyuanX: Spec says OpenAL1.1 is thread safe already
 //
 bool OpenALStream::Start()
 {
-  m_run_thread.Set();
-  bool bReturn = false;
-
-  ALDeviceList pDeviceList;
-  if (pDeviceList.GetNumDevices())
+  if (!palcIsExtensionPresent(nullptr, "ALC_ENUMERATION_EXT"))
   {
-    char* defDevName = pDeviceList.GetDeviceName(pDeviceList.GetDefaultDevice());
+    PanicAlertT("OpenAL: can't find sound devices");
+    return false;
+  }
 
-    INFO_LOG(AUDIO, "Found OpenAL device %s", defDevName);
+  const char* defaultDeviceName = palcGetString(nullptr, ALC_DEFAULT_DEVICE_SPECIFIER);
+  INFO_LOG(AUDIO, "Found OpenAL device %s", defaultDeviceName);
 
-    ALCdevice* pDevice = alcOpenDevice(defDevName);
-    if (pDevice)
-    {
-      ALCcontext* pContext = alcCreateContext(pDevice, nullptr);
-      if (pContext)
-      {
-        // Used to determine an appropriate period size (2x period = total buffer size)
-        // ALCint refresh;
-        // alcGetIntegerv(pDevice, ALC_REFRESH, 1, &refresh);
-        // period_size_in_millisec = 1000 / refresh;
-
-        alcMakeContextCurrent(pContext);
-        thread = std::thread(&OpenALStream::SoundLoop, this);
-        bReturn = true;
-      }
-      else
-      {
-        alcCloseDevice(pDevice);
-        PanicAlertT("OpenAL: can't create context for device %s", defDevName);
-      }
-    }
-    else
-    {
-      PanicAlertT("OpenAL: can't open device %s", defDevName);
-    }
+  ALCdevice* pDevice = palcOpenDevice(defaultDeviceName);
+  if (!pDevice)
+  {
+    PanicAlertT("OpenAL: can't open device %s", defaultDeviceName);
+    return false;
   }
-  else
+
+  ALCcontext* pContext = palcCreateContext(pDevice, nullptr);
+  if (!pContext)
   {
-    PanicAlertT("OpenAL: can't find sound devices");
+    palcCloseDevice(pDevice);
+    PanicAlertT("OpenAL: can't create context for device %s", defaultDeviceName);
+    return false;
   }
 
-  return bReturn;
+  // Used to determine an appropriate period size (2x period = total buffer size)
```

Comment:
I honestly didn't think about it when I was reorganizing this method, but yeah these can be removed.